### PR TITLE
Add DDS-CNF density method to DensityArchive

### LINKED ISSUE #545: [FEATURE REQUEST] Implement DDS-CNF
### Description

#487 implemented DDS-KDE. We would like to add DDS-CNF. There is some uncertainty here about how to clean it up and make sure that the imports are installed nicely, as DDS-CNF requires PyTorch and zuko.

### PR DESCRIPTION
**Pull request by efsiatras**
## Description

Adds `density_method="cnf"` to `DensityArchive` (closes #545), implementing the continuous normalizing flow variant of Density Descent Search from [Lee 2024](https://dl.acm.org/doi/10.1145/3638529.3654001) ("Density Descent for Diversity Optimization", GECCO 2024). This is the companion to DDS-KDE, which was added in #487.

The CNF is backed by `zuko.flows.CNF` and fine-tuned on the reservoir-sampled buffer at the end of each `add()` call, matching Algorithm 1 line 12 of the paper. Density queries return log-density (same semantics as `kde_sklearn`).

The main concern raised in the issue was clean optional-dep management. We handle this with a new `flows = ["torch", "zuko>=1.0.0"]` extra in pyproject.toml and lazy imports inside `_CNFDensityEstimator.__init__`. If torch or zuko is missing, the user gets a clear `ImportError` pointing to `pip install ribs[flows]`.

## TODO

- [x] Add `_CNFDensityEstimator` helper class
- [x] Extend `DensityArchive.__init__`, `compute_density`, and `add` with `density_method="cnf"` branch
- [x] New `flows` optional-dep group in `pyproject.toml`, added to `all`
- [x] Tests: initial zeros, first-add behavior, post-fit density, dtype (float32 + float64), min-buffer gating, functional correctness, input validation, reproducibility, small-buffer edge case, retraining, higher-dim (measure_dim=5), error paths
- [x] Rename existing "DDS" row to "DDS-KDE" and add "DDS-CNF" row in `docs/supported-algorithms.md`
- [x] Changelog entry in `HISTORY.md`

## Questions

- [x] Are the default CNF hyperparameters reasonable (`hidden_features=(64,64)`, Adam lr=1e-3, 100 train steps per `add()` call)? The paper doesn't specify architecture details so these are our choice; happy to adjust if you have preferences from your experiments.
- [x] Should `compute_density` return log-density (current, matching `kde_sklearn`) or raw density (matching `kde`)? Log-density is more numerically stable and `DensityRanker` works with either since it sorts ascending.

## Status

- [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md)
- [x] I have linted and formatted my code with `ruff` and `pylint`
- [x] I have tested my code by running `pytest`
- [x] I have added a description of my change to the changelog in `HISTORY.md`
- [x] This PR is ready to go

### CODE REVIEW

File: ribs/archives/_density_archive.py [hunk 1]
```diff
@@ -50,6 +55,181 @@ def gaussian_kde_measures(
...
+    ) -> None:
+        # Lazy import keeps torch/zuko optional. If either is missing, surface a
+        # clear install hint instead of a bare ModuleNotFoundError.
+        try:
+            import torch  # pylint: disable = import-outside-toplevel
+            import zuko  # pylint: disable = import-outside-toplevel
+        except ImportError as exc:
+            raise ImportError(_FLOWS_EXTRA_HINT) from exc
+
+        self._torch = torch
```
Comment (btjanaka) [inline ribs/archives/_density_archive.py:111]: One issue I see with setting the modules as attributes is that it's a bit more troublesome to save this density estimator object, as modules can't be pickled. I recommend taking a look at how I handled this in discount_models.py: https://github.com/icaros-usc/pyribs/blob/master/ribs/discount_models.py#L25

File: ribs/archives/_density_archive.py [hunk 2]
```diff
@@ -14,6 +14,11 @@
 from ribs.archives._utils import parse_all_dtypes
 from ribs.typing import BatchData, Float, Int
 
+_FLOWS_EXTRA_HINT = (
+    "DDS-CNF requires the 'flows' optional dependency group. "
```
Comment (btjanaka) [inline ribs/archives/_density_archive.py:18]: Update to just say to install torch and zuko.

**Review by btjanaka — CHANGES_REQUESTED** [review]
Hi @efsiatras, thanks for the PR! I will need to take a closer look at the flow matching code, but here are my comments on some (actionable) housekeeping things:

1. Regarding dependencies, I submitted #692, which adds the PyTorch dep to ribs[all]. Then, I modified this PR to just add zuko to ribs[all] in pyproject.toml. In general, I think we should only add extras to pyribs if they serve a broad purpose, since extras are a bit troublesome to add (consider the Conda package, for instance). Furthermore, the flows extra is only used for one setting of one component. Instead, I think we should just have a note in the documentation of DensityArchive to install zuko>=1.0.0 and torch. Then, we can also have the dependency check, which you implemented and I left some comments on. This is roughly what we have done for [PyCMAEvolutionStrategy](https://docs.pyribs.org/en/stable/api/ribs.emitters.opt.PyCMAEvolutionStrategy.html).

2. For tests, we usually put tests that require an extra dependency in their own folder, e.g., there's `tests/emitters_pycma`. Could you move the tests you added to `tests/archives_zuko/density_archive_test.py`? I have already added a CI step that will call these tests once they are there. This way, we can guarantee that the tests are called in CI; we actually don't want dependency checks in the tests as that puts holes in our test suite.

Regarding your questions:
 
1. For hyperparameters, I searched up our code for the DDS paper, and it seems we had:
    - AdamW (lr=1e-3) -- doubt Adam vs AdamW makes a huge difference though
    - 20 train steps per add call
    - 3 hidden layers of 256 for the architecture
   Could you try those out and see if there's any performance difference? If there is no difference, I'm happy with the ones you mentioned as defaults. If you have time, could you use the benchmark in sphere.py, similar to what I requested in #690?

2. Regarding `compute_density`, I think log density is alright. The ranker should indeed work fine with either.

File: tests/archives/density_archive_test.py [hunk 3]
```diff
@@ -1,11 +1,25 @@
...
+import importlib.util
+
 import numpy as np
 import pytest
 from numpy.testing import assert_allclose
 
+import ribs.archives._density_archive as _density_archive_module
 from ribs.archives import DensityArchive
 
+# DDS-CNF tests below require torch + zuko (the `flows` optional extra). We
```
Comment (btjanaka) [inline tests/archives/density_archive_test.py:12]: No need for these dependency checks; see main review comment.

**Comment by btjanaka** [comment]
@efsiatras Btw, I gave you write access to the repo. Let me know if you received it!

File: ribs/archives/_density_archive.py [hunk 4]
```diff
@@ -197,6 +419,16 @@ def compute_density(self, measures: ArrayLike) -> np.ndarray:
                 **self._sklearn_kwargs,
             ).fit(self.buffer)
             return kde.score_samples(measures).astype(self._measures_dtype)
+        elif self._density_method == "cnf":
+            # The CNF returns log-density (this matches kde_sklearn's
+            # semantics). Before the estimator has been fit -- i.e., during the
+            # first add() call or before the buffer crosses
+            # `cnf_min_buffer_size` -- this returns zeros, mirroring how KDE
+            # returns zeros on an empty buffer.
+            assert self._cnf_estimator is not None
```
Comment (btjanaka) [inline ribs/archives/_density_archive.py]: Can you replace the assertions with RuntimeError or another exception? I avoid assertions in pyribs because they don't provide a lot of information to the end user.

File: ribs/archives/_density_archive.py [hunk 5]
```diff
@@ -282,4 +516,14 @@ def add(
...
             n_remaining -= skip
             self._n_skip -= skip
+
+        # For DDS-CNF, Algorithm 1 line 12 of Lee et al. 2024 calls for the
+        # density estimator to be refit on the updated buffer at the end of
+        # every iteration. We fine-tune the CNF here so that the next
+        # compute_density() call uses an up-to-date flow. KDE methods do not
+        # need this step because they are non-parametric.
+        if self._density_method == "cnf":
+            assert self._cnf_estimator is not None
```
Comment (btjanaka) [inline ribs/archives/_density_archive.py]: Same here Re: assertions.

File: ribs/archives/_density_archive.py (hunk 4 above)
Comment (btjanaka) [reply ribs/archives/_density_archive.py]: Edit: I think we can just remove the assert entirely.

**Comment by btjanaka** [comment]
TODO (as of 5/9/26):
- [x] Review `_CNFDensityEstimator`

Followup PR: #707

**Review by btjanaka — APPROVED** [review]
Hi @efsiatras, I had some spare time on my hands, so I went ahead and made all the fixes to the PR, as I also want to get it out for v0.11.0. Once again, thank you so much for contributing this!
